DeviceBridge/Services/ConnectionStatusSubscriptionService.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using DeviceBridge.Models;
+using DeviceBridge.Providers;
+using NLog;
+
+namespace DeviceBridge.Services
+{
+    /// <summary>
+    /// CRUD operations for connection status subscriptions. It synchronizes the DB and device client callback update to make sure
+    /// that the registered callback always reflects the actual subscription stored in the DB.
+    /// The synchronization is separate from data subscriptions, which might take a long time to synchronize due to connection creation.
+    /// </summary>
+    public class ConnectionStatusSubscriptionService : IConnectionStatusSubscriptionService
+    {
+        private readonly IStorageProvider _storageProvider;
+        private readonly IConnectionManager _connectionManager;
+        private readonly ISubscriptionCallbackFactory _subscriptionCallbackFactory;
+        private readonly Logger _logger;
+
+        private ConcurrentDictionary<string, SemaphoreSlim> _connectionStatusSubscriptionSyncSemaphores = new ConcurrentDictionary<string, SemaphoreSlim>();
+
+        public ConnectionStatusSubscriptionService(Logger logger, IConnectionManager connectionManager, IStorageProvider storageProvider, ISubscriptionCallbackFactory subscriptionCallbackFactory)
+        {
+            _logger = logger;
+            _storageProvider = storageProvider;
+            _connectionManager = connectionManager;
+            _subscriptionCallbackFactory = subscriptionCallbackFactory;
+        }
+
+        public async Task<DeviceSubscription> GetConnectionStatusSubscription(Logger logger, string deviceId, CancellationToken cancellationToken)
+        {
+            return await _storageProvider.GetDeviceSubscription(logger, deviceId, DeviceSubscriptionType.ConnectionStatus, cancellationToken);
+        }
+
+        public async Task<DeviceSubscription> CreateOrUpdateConnectionStatusSubscription(Logger logger, string deviceId, string callbackUrl, CancellationToken cancellationToken)
+        {
+            var mutex = _connectionStatusSubscriptionSyncSemaphores.GetOrAdd(deviceId, new SemaphoreSlim(1, 1));
+            await mutex.WaitAsync();
```

Comment:
If there is multiple service instances, is that possible this operation for same device id falls in different service instance?

Reply:
Do we need a distribute lock here?

Reply:
This service is initialized as a .NET singleton with `AddSingleton` on startup, so only one instance exists at a time. It's ok if one instance is deleted and recreated, as long as two service instances are not accepting requests at once.
The other services also rely on being singletons.

Reply:
I mean deployed microservice instances in ACI.

Reply:
Oh, this entire architecture assumes that we only have one container running. Many other things will go wrong if we have multiple instances (e.g., Hub can't accept connections for the same device coming from multiple places) :)